library/objective-c/EnvoyEngine.mm

```diff
@@ -1,16 +1,157 @@
 #import "library/objective-c/EnvoyEngine.h"
+#import "library/objective-c/EnvoyTypes.h"
 
 #import "library/common/include/c_types.h"
 #import "library/common/main_interface.h"
 
+#pragma mark - utility functions to move elsewhere
+static envoy_headers toNativeHeaders(EnvoyHeaders *headers) {
+  int length = 0;
+  for (id headerList in headers) {
+    length += headerList.count;
+  }
+  envoy_header *header_array = (envoy_header *)malloc(sizeof(envoy_header) * length);
+  int j = 0;
+  for (id headerKey in headers) {
+    NSArray *headerList = headers[headerKey];
+    for (id headerValue in headerList) {
+      header_array[j++] = {
+        toManagedNativeString(headerKey),
+        toManagedNativeString(headerValue)
+      };
+    }
+  }
+  return { length, header_array };
+}
+
+
+
+static envoy_data toUnmanagedNativeData(NSData *data) {
+  // TODO: implement me
+}
+
+static envoy_data toManagedNativeString(NSString *s) {
+  int length = s.length;
+  char *native_string = (char *)malloc(sizeof(char) * length);
+  memcpy(s.UTF8String, native_string, length)
+  return {length, native_string, ios_free_native_data, native_string};
+}
+
+static void ios_free_native_data(void *context) {
+  free(context);
+}
+
+static void ios_release_native_data(void *context) {
+  // TODO: implement me
+}
+
+static NSData * to_ios_data(envoy_data data) {
+  // TODO: investigate buffer ownership
+  // Possibly extend/subclass NSData to call envoy_data.release on dealloc and have release drain the underlying Envoy buffer instance
+  // FIXME: do we need to autorlease here?
+  return [[NSData dataWithBytesNoCopy:data.bytes length:data.length] autorelease];
+}
+
+static EnvoyHeaders * to_ios_headers(envoy_headers headers) {
+  NSMutableDictianary *headerDict = [NSMutableDictionary new];
+  for (uint_fast32_t i = 0; i < headers.length; i++) {
```

Comment:
Why do we use `int`, `uint64_t` and `uint_fast32_t` in different loops?

Reply:
moving fast, changing stuff.